### PR TITLE
[Breaking change] Remove `createFix`, and just use the Replacement(s) as the fix.

### PR DESCRIPTION
**Pull request by andy-hanson**
#### PR checklist

- [ ] Addresses an existing issue: #0000
- [ ] New feature, bugfix, or enhancement
  - [ ] Includes tests
- [ ] Documentation update

#### Overview of change:

The `Fix` class is not really necessary. It's easier to just pass parameters to `addFailure`.
This could be viewed as a continuation of #1845. So now instead of `this.addFailure(this.createFailure(..., this.createFix(foo)))`, you just `this.addFailure(..., foo)`.

#### CHANGELOG.md entry:

[api] Remove `createFix`. Replacements should be passed directly into `addFailure`.

### CODE REVIEW

File: src/language/rule/rule.ts [hunk 1]
```diff
@@ -188,19 +192,10 @@ export class Fix {
...
     }
 
-    constructor(private innerRuleName: string, private innerReplacements: Replacement[]) {
-    }
-
-    get ruleName() {
-        return this.innerRuleName;
-    }
-
-    get replacements() {
```
Comment (andy-hanson) [inline src/language/rule/rule.ts]: `get innerReplacements() { return this.innerInnerReplacements; }` 😜

File: src/language/rule/rule.ts [hunk 2]
```diff
@@ -138,6 +138,10 @@ export class Replacement {
         return replacements.reduce((text, r) => r.apply(text), content);
     }
 
+    public static replaceNode(node: ts.Node, text: string): Replacement {
```
Comment (ajafff) [inline src/language/rule/rule.ts]: please add an optional `sourceFile` parameter, that will be passed to `node.getStart` (for performance)

File: src/language/rule/rule.ts [hunk 3]
```diff
@@ -240,19 +235,21 @@ export class RuleFailure {
...
                 private ruleName: string,
-                private fix?: Fix) {
+                fix?: Replacement | Replacement[]) {
 
         this.fileName = sourceFile.fileName;
         this.startPosition = this.createFailurePosition(start);
         this.endPosition = this.createFailurePosition(end);
         this.rawLines = sourceFile.text;
         this.ruleSeverity = "error";
+        this.fix = fix ? new Fix(Array.isArray(fix) ? fix : [fix]) : undefined;
```
Comment (ajafff) [inline src/language/rule/rule.ts]: use `arrayify`?

File: src/language/walker/ruleWalker.ts [hunk 4]
```diff
@@ -65,7 +65,7 @@ export class RuleWalker extends SyntaxWalker implements IWalker {
     }
 
     /** @deprecated Prefer `addFailureAt` and its variants. */
-    public createFailure(start: number, width: number, failure: string, fix?: Fix): RuleFailure {
+    public createFailure(start: number, width: number, failure: string, fix?: Replacement | Replacement[]): RuleFailure {
```
Comment (ajafff) [inline src/language/walker/ruleWalker.ts]: I would prefer a rest parameter like `createFix` had

File: src/language/walker/walkContext.ts [hunk 5]
```diff
@@ -17,31 +17,27 @@
...
     }
 
-    public addFailure(start: number, end: number, failure: string, fix?: Fix) {
+    public addFailure(start: number, end: number, failure: string, fix?: Replacement | Replacement[]) {
         const fileLength = this.sourceFile.end;
-        this.failures.push(
-            new RuleFailure(this.sourceFile, Math.min(start, fileLength), Math.min(end, fileLength), failure, this.ruleName, fix),
-        );
+        start = Math.min(start, fileLength);
+        end = Math.min(end, fileLength);
```
Comment (ajafff) [inline src/language/walker/walkContext.ts]: Since you are already changing the style of the code, you could inline `this.sourceFile.end` here and remove `fileLength`

File: src/language/walker/ruleWalker.ts (hunk 4 above)
Comment (andy-hanson) [inline src/language/walker/ruleWalker.ts]: I tried that at first, but that can make life difficult if you have a function returning optional fixes. You would need to use `const fix = optionalFix(); ctx.addFailureAtNode(node, failure, ...(fix ? fix : []))`, instead of just `ctx.addFailureAtNode(node, failure, optionalFix())`.

File: src/language/rule/rule.ts [hunk 6]
```diff
@@ -240,19 +212,21 @@ export class RuleFailure {
...
     private ruleSeverity: RuleSeverity;
+    private fix?: Fix;
 
     constructor(private sourceFile: ts.SourceFile,
                 start: number,
                 end: number,
                 private failure: string,
                 private ruleName: string,
-                private fix?: Fix) {
+                fix?: Fix) {
```
Comment (adidahiya) [inline src/language/rule/rule.ts]: why this change?

File: src/rules/alignRule.ts [hunk 7]
```diff
@@ -108,13 +108,9 @@ class AlignWalker extends Lint.RuleWalker {
...
             if (curPos.line !== prevPos.line && curPos.character !== alignToColumn) {
                 const diff = alignToColumn - curPos.character;
-                let fix;
-                if (0 < diff) {
-                    fix = this.createFix(this.appendText(node.getStart(), " ".repeat(diff)));
-                } else {
-                    fix = this.createFix(this.deleteText(node.getStart() + diff, -diff));
-                }
-                this.addFailureAtNode(node, kind + Rule.FAILURE_STRING_SUFFIX, fix);
+                this.addFailureAtNode(node, kind + Rule.FAILURE_STRING_SUFFIX, 0 < diff
```
Comment (adidahiya) [inline src/rules/alignRule.ts]: IMO the inlined `fix` variable is hard to read, can you save it a variable before calling this method?

**Comment by andy-hanson** [comment]
#2403 merged, so closing.
